sw/device/lib/otbn.h

```diff
@@ -0,0 +1,248 @@
+ * structure.
+ */
+typedef struct otbn_ptr {
+  /**
+   * Address of the symbol in normal CPU address space.
+   */
+  const char *ptr_in_cpu_addr_space;
+} otbn_ptr_t;
+
+/**
+ * The result of an OTBN operation.
+ */
+typedef enum otbn_result {
+  /**
+   * The operation succeeded.
+   */
+  kOtbnOk = 0,
+  /**
+   * An unspecified failure occurred.
+   */
+  kOtbnError = 1,
+  /**
+   * The precondition check for an argument passed into the function failed.
+   */
+  kOtbnBadArg = 2,
+  /**
+   * The chosen application is not part of the loaded application.
+   */
+  kOtbnFuncNotInApp = 3,
+  /**
+   * No application is loaded.
+   */
+  kOtbnAppNotLoaded = 4,
+  /**
+   * The execution of the application on OTBN failed.
+   *
+   * More specific error information can be obtained with
+   * `dif_otbn_get_err_code()`.
+   */
+  kOtbnExecutionFailed = 5,
+} otbn_result_t;
+
+/**
+ * OTBN context structure.
+ *
+ * Use `otbn_init()` to initialize.
+ */
+typedef struct otbn {
+  /**
+   * The OTBN DIF to access the hardware.
+   */
+  dif_otbn_t dif;
+
+  /**
+   * The application loaded or to be loaded into OTBN.
+   */
+  const otbn_app_t *app;
+
+  /**
+   * Is the application loaded into OTBN?
+   */
+  bool app_is_loaded;
+} otbn_t;
+
+/**
+ * Make an embedded OTBN application image available for use.
+ *
+ * Make symbols available that indicate the start and the end of instruction
+ * and data memory regions, as they are stored in the device memory.
+ *
+ * @param app_name name of the application to load, which is typically the
+ *                 name of the main (assembly) source file.
+ */
+#define OTBN_USE_APP(app_name)                            \
+  extern const char _otbn_app_##app_name##__imem_start[]; \
+  extern const char _otbn_app_##app_name##__imem_end[];   \
+  extern const char _otbn_app_##app_name##__dmem_start[]; \
+  extern const char _otbn_app_##app_name##__dmem_end[];
+
+/**
+ * Initialize the OTBN application information structure
+ *
+ * After making all required symbols from the application image available
+ * through `OTBN_USE_APP()`, use this macro to initialize an `otbn_app_t`
+ * struct with those symbols.
+ *
+ * @param app_name name of the application to load
+ * @see OTBN_USE_APP()
+ */
+#define OTBN_APP_T_INIT(app_name)                     \
+  {                                                   \
+    .imem_start = _otbn_app_##app_name##__imem_start, \
+    .imem_end = _otbn_app_##app_name##__imem_end,     \
+    .dmem_start = _otbn_app_##app_name##__dmem_start, \
+    .dmem_end = _otbn_app_##app_name##__dmem_end,     \
+  }
+
+/**
+ * Make a symbol in the OTBN application image available.
+ *
+ * Symbols are typically function or data pointers, i.e. labels in assembly
+ * code.
+ *
+ * @param app_name    name of the application the function is contained in
+ * @param symbol_name name of the symbol (function, label)
+ */
+#define OTBN_USE_PTR(app_name, symbol_name) \
```

Comment:
I'm not sure that this macro adds a ton of value. Why not simply document that the symbols have a specific name, and let users pull them up how they see fit? I think it makes some sense for the imem and dmem symbols, since those seem more boilerplatey.

(Though, thinking about my other comment about unifying the two above macros, maybe it makes sense to do that here, too.)

Reply:
Two reasons, both are effectively a matter of taste.

- It's nice to have a symmetry between the app and the pointer usage and hide complexities around the use of the symbol and their naming.
- I'd like to avoid having too many sites where the real symbol name is used. The symbol name is generated by tooling and pre-processing, and I wouldn't be surprised if we change the naming there as we move along. Having only a single site where we need to change names is helpful.

Reply:
Added a sentence "Use this macro instead of manually declaring the symbols as symbol names
might change." to the documentation of the macros, and renamed `OTBN_USE_PTR()` and `OTBN_USE_APP()` to `OTBN_DECLARE_PTR_SYMBOL()` and `OTBN_DECLARE_APP_SYMBOLS()`.